src/pudl/analysis/allocate_net_gen.py

```diff
@@ -324,183 +509,437 @@ def stack_generators(
-    )
-
-    gen_assoc = (
-        pd.merge(
-            gen_assoc,
-            gen_assoc.groupby(by=IDX_ESC)[["capacity_mw", "net_generation_mwh_g_tbl"]]
-            .sum(min_count=1)
-            .add_suffix("_fuel")
-            .reset_index(),
-            on=IDX_ESC,
+        .pipe(remove_inactive_generators)
+        .pipe(
+            _allocate_unassociated_records,
+            idx_cols=IDX_PM_ESC,
+            col_w_unexpected_codes="prime_mover_code",
+            data_columns=[f"{col}_gf_tbl" for col in DATA_COLUMNS],
         )
-        .pipe(apply_pudl_dtypes, "eia")
-        .pipe(_associate_unconnected_records)
+        .drop(columns=["_merge"])  # drop do we can do this again in the bf_summed merge
+        .merge(
+            bf_summed, on=IDX_GENS_PM_ESC, how="outer", validate="m:1", indicator=True
+        )
+        .pipe(
+            _allocate_unassociated_records,
+            idx_cols=IDX_GENS_PM_ESC,
+            col_w_unexpected_codes="energy_source_code",
+            data_columns=["fuel_consumed_mmbtu_bf_tbl"],
+        )
+        .drop(columns=["_merge"])
     )
+
+    # replace zeros with small number to avoid div by zero errors when calculating allocation fraction
+    data_columns = [
+        "fuel_consumed_mmbtu_gf_tbl",
+        "fuel_consumed_for_electricity_mmbtu_gf_tbl",
+        "net_generation_mwh_gf_tbl",
+        "fuel_consumed_mmbtu_bf_tbl",
+    ]
+    gen_assoc[data_columns] = gen_assoc[data_columns].replace(0, MISSING_SENTINEL)
+
+    # calculate the total capacity in every fuel group
+    gen_assoc = pd.merge(
+        gen_assoc,
+        gen_assoc.groupby(by=IDX_ESC)[["capacity_mw", "net_generation_mwh_g_tbl"]]
+        .sum(min_count=1)
+        .add_suffix("_fuel")
+        .reset_index(),
+        on=IDX_ESC,
+        how="outer",
+    ).pipe(apply_pudl_dtypes, "eia")
     return gen_assoc
 
 
-def remove_retired_generators(gen_assoc):
+def remove_inactive_generators(gen_assoc: pd.DataFrame) -> pd.DataFrame:
     """Remove the retired generators.
 
     We don't want to associate net generation to generators that are retired
-    (or proposed! or any other `operational_status` besides `existing`).
-
-    We do want to keep the generators that retire mid-year and have generator
-    specific data from the generation_eia923 table. Removing the generators
-    that retire mid-report year and don't report to the generation_eia923 table
-    is not exactly a great assumption. For now, we are removing them. We should
-    employ a strategy that allocates only a portion of the generation to them
-    based on their operational months (or by doing the allocation on a monthly
-    basis).
+    (or proposed! or any other ``operational_status`` besides ``existing``).
+
+    We do want to keep the generators that report operational statuses other
+    than ``existing`` but which report non-zero data despite being ``retired`` or
+    ``proposed``. This includes several categories of generators/plants:
```

Comment:
omfg this is so complicated.  It seems to directly contradict what you just said in the previous sentence which is jarring and makes me wonder if the documentation is conflicting with itself. Some kind of foreshadowing would help. Maybe "We don't want to allocate generation or fuel consumption to generators that are truly no longer operational. However, ..."